src/commands/txCommands/pullAll.js

```diff
@@ -0,0 +1,73 @@
+const chalk = require('chalk');
+const child_process = require('child_process');
+const Rx = require('rxjs');
+const txlib = require('./util');
+const txPullResource = require('./util/pullResourceTrns');
+
+const getProjectResourcesList$ =
+	txlib.resources$
+		.flatMap(Rx.Observable.from)
+		.reduce((resources, resource) => [ ...resources, resource ], []);
+
+/**
+ * recursively goes through list of resources and pulls trn content down
```

Comment:
seems odd to me to use recursion rather than just iterating over a list - Observables are really good at performing operations on lists in a flat, non-recursive way, so I think this could be refactored to be more clear

```js
const child_process$ = Observable.fromNodeCallback(child_process.exec);
Observable.from(resources)
  .flatMap(resource =>
    return txPullResource.pullResourceContent$
      .flatMap( => {
        if (!commitOnComplete) {
          return Observable.empty();
        }
        return child_process$('git status')
          .flatMap(([stdout, stderr]) => {
            if (stdout.includes('nothing to commit')) {
              console.log(chalk.yellow(`- no changes to commit for '${resource}'`));
              return Observable.empty();
            }
            child_process.execSync('git add src/trns');
            return child_process$(`git commit -m "tx:pull for ${resource.replace(/-/g, '_')}" --no-verify`)
              .do(() => console.log(chalk.yellow(`- no changes to commit for '${resources[iteration]}'`)));
      });
  });
```

something like that - probably some more optimizations possible and you can break out the nested functions into pure functions defined at the top level of the module...

But generally, probably better to avoid the complexity of recursion when you're operating on a list of a fixed length.

Reply:
First pass at the refactor, but this will run the trn pull in sequence wanted to get your your take @mmcgahan before i keep going...

```js
        const getProjectResourcesList$ =
            txlib.resources$
                .flatMap(Rx.Observable.from);

        getProjectResourcesList$
            .map(
                resource =>
                    Rx.Observable.create(observer => {
                        pullResourceTrns.pullResourceContent$(resource)
                            .subscribe(null, null, () => {
                                console.log(chalk.green('done...', resource));
                                observer.complete();
                            });
                    })
            )
            .concatAll()
            .subscribe(null, null, () => console.log(chalk.green('pulls done...')));
```

Reply:
i think this:
```js
            .map(
                resource =>
                    Rx.Observable.create(observer => {
                        pullResourceTrns.pullResourceContent$(resource)
                            .subscribe(null, null, () => {
                                console.log(chalk.green('done...', resource));
                                observer.complete();
                            });
                    })
            )
```

is equivalent to this (without logging)

```js
.map(pullResourceTrns.pullResourceContent$)
```

if you want logging:

```js
.map(r => pullResourceTrns.pullResourceContent$(r).do(null, null, () => console.log(...)))
```


but I think instead of `concatAll()`, you might want to use `flatMap` with a concurrency of `1`

```js
.flatMap(pullResourceTrns.pullResourceContent$, 1).subscribe(...)
```

Another option that you might want to consider is turning all these observables into Promises so that they are easier to reason about - I think when the web platform takes over TRN development, we'll start migrating this away from observables, which have made the script much more difficult to debug and maintain...

Reply:
Ahhh `flatMap(thing, 1)` did it. i thought there was another param between there. 🎉